### PR TITLE
[SPARK-31182][CORE][ML] PairRDD support aggregateByKeyWithinPartitions

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
1, impl `aggregateByKeyWithinPartitions` and `reduceByKeyWithinPartitions`
2, use `aggregateByKeyWithinPartitions` in `RobustScaler`

### Why are the changes needed?
When implementing `RobustScaler`, I was looking for a way to guarantee that the `QuantileSummaries` in `aggregateByKey` are compressed at the _map_ side.

(before `merge` and `qurey`, the `QuantileSummaries` must be compressed)

Then I only found a tricky method to work around (yet not applied), and there was no method for this.
previous discussions were [here](https://github.com/apache/spark/pull/26803)

### Does this PR introduce any user-facing change?
Yes, add new methods for PairRDD

### How was this patch tested?
added testsuites and existing ones

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala [hunk 1]
```diff
@@ -99,6 +99,35 @@ class PairRDDFunctions[K, V](self: RDD[(K, V)])
...
+   * Users provide three functions:
+   *
+   *  - `createCombiner`, which turns a V into a C (e.g., creates a one-element list)
+   *  - `mergeValue`, to merge a V into a C (e.g., adds it to the end of a list)
+   *  - `mergeCombiners`, to combine two C's into a single one.
+   *
+   * @note V and C can be different -- for example, one might group an RDD of type
+   * (Int, Int) into an RDD of type (Int, Seq[Int]).
+   */
+  def combineByKeyWithClassTagWithinPartitions[C](
```
Comment (zhengruifeng) [inline core/src/main/scala/org/apache/spark/rdd/PairRDDFunctions.scala]: this impl follows `combineByKeyWithClassTag` (treat as if `self.partitioner == Some(partitioner)`)
```scala
def combineByKeyWithClassTag[C](
      createCombiner: V => C,
      mergeValue: (C, V) => C,
      mergeCombiners: (C, C) => C,
      partitioner: Partitioner,
      mapSideCombine: Boolean = true,
      serializer: Serializer = null)(implicit ct: ClassTag[C]): RDD[(K, C)] = self.withScope {
    require(mergeCombiners != null, "mergeCombiners must be defined") // required as of Spark 0.9.0
    if (keyClass.isArray) {
      if (mapSideCombine) {
        throw new SparkException("Cannot use map-side combining with array keys.")
      }
      if (partitioner.isInstanceOf[HashPartitioner]) {
        throw new SparkException("HashPartitioner cannot partition array keys.")
      }
    }
    val aggregator = new Aggregator[K, V, C](
      self.context.clean(createCombiner),
      self.context.clean(mergeValue),
      self.context.clean(mergeCombiners))
    if (self.partitioner == Some(partitioner)) {
      self.mapPartitions(iter => {
        val context = TaskContext.get()
        new InterruptibleIterator(context, aggregator.combineValuesByKey(iter, context))
      }, preservesPartitioning = true)
    } else {
      new ShuffledRDD[K, V, C](self, partitioner)
        .setSerializer(serializer)
        .setAggregator(aggregator)
        .setMapSideCombine(mapSideCombine)
    }
  }
```

File: mllib/src/main/scala/org/apache/spark/ml/feature/RobustScaler.scala [hunk 2]
```diff
@@ -204,17 +204,18 @@ object RobustScaler extends DefaultParamsReadable[RobustScaler] {
...
+        vec.iterator.filterNot(_._2.isNaN)
+      }.aggregateByKeyWithinPartitions(
         new QuantileSummaries(QuantileSummaries.defaultCompressThreshold, relativeError))(
         seqOp = (s, v) => s.insert(v),
         combOp = (s1, s2) => s1.compress.merge(s2.compress)
-      ).map { case ((_, i), s) => (i, s)
-      }.reduceByKey { case (s1, s2) => s1.compress.merge(s2.compress) }
+      ).mapPartitionsWithIndex { case (pid, iter) =>
+        val p = pid % scale
+        iter.map { case (col, s) => ((p, col), s.compress) }
```
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RobustScaler.scala]: here trigger compression at the _map_ side

File: mllib/src/main/scala/org/apache/spark/ml/feature/RobustScaler.scala [hunk 3]
```diff
@@ -204,17 +204,18 @@ object RobustScaler extends DefaultParamsReadable[RobustScaler] {
...
+      }.reduceByKey { case (s1, s2) => s1.merge(s2)
```
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/ml/feature/RobustScaler.scala]: then no longer to confirm compression like `s1.compress.merge(s2.compress)`

**Comment by zhengruifeng** [comment]
friendly ping @srowen

**Comment by wetneb** [comment]
Naive question from a newbie: you are introducing new methods for PairRDDs, so would it make sense to also expose them in the Java API (as methods of `JavaPairRDD`)? Or is it generally expected that the Java API is updated separately?

**Comment by srowen** [comment]
Well, I think that's part of the issue here - if it's public you kind of need to support it everywhere and for a long time. I don't know if it's worth it but I've lost the thread on this PR and would have to recall the motivation.

**Comment by zhengruifeng** [comment]
I tend to close it, since I can always workaround it. Maybe it is not necessary.

**Pull request closed without merging**